addons/product/models/product_template.py

```diff
@@ -967,24 +1023,24 @@ def _get_possible_combinations(self, parent_combination=None, necessary_values=N
 
         necessary_values = necessary_values or self.env['product.template.attribute.value']
         necessary_attributes = necessary_values.mapped('attribute_id')
-        ptal_stack = [self.valid_product_template_attribute_line_ids.filtered(lambda ptal: ptal.attribute_id not in necessary_attributes)]
-        combination_stack = [necessary_values]
-
-        # keep going while we have attribute lines to test
-        while len(ptal_stack):
-            attribute_lines = ptal_stack.pop()
-            combination = combination_stack.pop()
-
-            if not attribute_lines:
-                # full combination, if it's possible return it, otherwise skip it
-                if self._is_combination_possible(combination, parent_combination):
-                    yield(combination)
-            else:
-                # we have remaining attribute lines to consider
-                for ptav in reversed(attribute_lines[0].product_template_value_ids):
-                    ptal_stack.append(attribute_lines[1:])
-                    combination_stack.append(combination + ptav)
 
+        attributes = self.valid_product_template_attribute_line_ids.filtered(lambda ptal: ptal.attribute_id not in necessary_attributes)
+
+        if not attributes and self._is_combination_possible(necessary_values, parent_combination):
+            yield necessary_values
+
+        # compute all possible values for each attribute
+        attributes_values = [
+            attribute.product_template_value_ids
+            for attribute in attributes
+        ]
+
+        for product in self._cartesian_product(attributes_values):
```

Comment:
what is product? Shouldn't it be `combination`?

Reply:
If it's not a full `combination`, maybe it should be called `partial_combination`

Reply:
product as the result of a multiplication, yes, I'll use `partial_combination`

Reply:
Yes I understand now that you just the mathematical language instead of the actual business names, which is why it was very confusing, in particular using the word product in this context.